kernel/kernel-api/src/test/scala/io/delta/kernel/internal/types/DataTypeJsonSerDeSuite.scala

```diff
@@ -245,6 +274,249 @@ object DataTypeJsonSerDeSuite {
+        .add("a2", new StructType()
+          .add("b1", new ArrayType(new StringType("SPARK.UTF8_LCASE"), false))
+          .add("b2", new MapType(
+            new StringType("ICU.UNICODE_CI"), new StringType("SPARK.UTF8_LCASE"), true), false)
+          .add("b3", new ArrayType(StringType.STRING, false))
+          .add("b4", new MapType(
+            StringType.STRING, StringType.STRING, false), false), true)
+        .add("a3", new StructType()
+          .add("b1", StringType.STRING, false)
+          .add("b2", new ArrayType(IntegerType.INTEGER, false), true), false)
+    ),
+    (
+      structTypeJson(Seq(
+        structFieldJson("a1", "\"string\"", true),
+        structFieldJson("a2", structTypeJson(Seq(
+          structFieldJson("b1", mapTypeJson(
+            arrayTypeJson(arrayTypeJson("\"string\"", true), true),
+            structTypeJson(Seq(
+              structFieldJson("c1", "\"string\"", false,
+                metadataJson = Some(
+                  s"""{"$COLLATIONS_METADATA_KEY"
+                     | : {"c1" : "SPARK.UTF8_LCASE"}}""".stripMargin)),
+              structFieldJson("c2", "\"string\"", true,
+                metadataJson = Some(
+                  s"""{"$COLLATIONS_METADATA_KEY"
+                     | : {\"c2\" : \"ICU.UNICODE\"}}""".stripMargin)),
+              structFieldJson("c3", "\"string\"", true))), true), true,
+            metadataJson = Some(
+              s"""{"$COLLATIONS_METADATA_KEY"
+               | : {"b1.key.element.element" : "SPARK.UTF8_LCASE"}}""".stripMargin)),
+          structFieldJson("b2", "\"long\"", true))), true),
+        structFieldJson("a3", arrayTypeJson(
+          mapTypeJson(
+            "\"string\"",
+            structTypeJson(Seq(
+              structFieldJson("b1", "\"string\"", false,
+                metadataJson = Some(
+                  s"""{"$COLLATIONS_METADATA_KEY"
+                     | : {"b1" : "SPARK.UTF8_LCASE"}}""".stripMargin)))),
+            false), false), true,
+          metadataJson = Some(
+            s"""{"$COLLATIONS_METADATA_KEY"
+               | : {"a3.element.key" : "ICU.UNICODE_CI"}}""".stripMargin)))),
+      new StructType()
+        .add("a1", StringType.STRING, true)
+        .add("a2", new StructType()
+          .add("b1", new MapType(
+            new ArrayType(
+              new ArrayType(
+                new StringType("SPARK.UTF8_LCASE"), true), true),
+            new StructType()
+              .add("c1", new StringType("SPARK.UTF8_LCASE"), false)
+              .add("c2", new StringType("ICU.UNICODE"), true)
+              .add("c3", StringType.STRING), true))
+          .add("b2", LongType.LONG), true)
+        .add("a3", new ArrayType(
+          new MapType(
+            new StringType("ICU.UNICODE_CI"),
+            new StructType()
+              .add("b1", new StringType("SPARK.UTF8_LCASE"), false), false), false), true)
+    )
+  )
+
+  val SAMPLE_JSON_TO_TYPES_WITH_COLLATION_DIFFERING = Seq(
```

Comment:
what does DIFFERING mean?

Reply:
idea of this test was to have just difference in StringType collation, i will make some better name